docs/src/main/mdoc/index.md

```diff
@@ -95,16 +97,100 @@ def fetchStatusInefficiently[F[_]: ConcurrentEffect](uri: Uri): F[Status] =
 
 @@@
 
-## Shutdown
+### Shutdown
 
 Clients created with this back end do not need to be shut down.
 
-## Further reading
+### Further reading
 
 For more details on the http4s-client, please see the [core client
-documentation][client].
+documentation][http4s-client].
+
+## WebSocket client
+
+This package also contains a functional WebSocket client. Please note that
+the API may change in the future.
+
+### Creation
+
+A `WebSocketClient` is created
+using an `HttpClient` as above. It is encouraged to use the same `HttpClient`
+to construct a `Client[F]` and a `WebSocketClient[F]`.
+
+```scala mdoc
+import org.http4s.client.jdkhttpclient._
+
+val (http, webSocket) =
+  IO(HttpClient.newHttpClient())
+    .map { httpClient =>
+      (JdkHttpClient[IO](httpClient), JdkWebSocketClient[IO](httpClient))
+    }
+    .unsafeRunSync()
+```
+
+If you do not need an HTTP client, you can also call `JdkWebSocketClient.simple[IO]` as above.
+
+### Overview
+
+We have the following WebSocket frame hierarchy:
+
+ - `WSFrame`
+   - `WSControlFrame`
+     - `WSFrame.Close`
+     - `WSFrame.Ping`
+     - `WSFrame.Pong`
+   - `WSDataFrame`
+     - `WSFrame.Text`
+     - `WSFrame.Binary`
+
+There are two connection modes: "low-level" and "high-level". Both manage the lifetime of a
+websocket connection via a [`Resource`][Resource].
+In the low-level mode, you can send and have to receive arbitrary `WSFrame`s.
+The high-level mode does the following things for you:
+
+ - Hides the control frames (you can still send Ping and Close frames).
+ - Responds to Ping frames with Pongs and echoes Close frames (the received Close frame is exposed
+   as a [`TryableDeferred`][TryableDeferred]).
+ - Groups the data frames by their `last` attribute.
+
+### Usage example
+
+We use the "high-level" connection mode to build a simple websocket app.
+
+```scala mdoc
+webSocket
+  .connectHighLevel(WSRequest(uri"wss://echo.websocket.org"))
+  .use { conn =>
+    for {
+      // send a single Text frame
+      _ <- conn.send(WSFrame.Text("reality"))
+      // send multiple frames (both Text and Binary are possible)
+      // "faster" than individual `send` calls
+      _ <- conn.sendMany(List(
+        WSFrame.Text("is often"),
+        WSFrame.Text("disappointing.")
+      ))
+      received <- conn
+        // a backpressured stream of incoming frames
+        .receiveStream
+        // we do not care about Binary frames (and will never receive any)
+        .collect { case WSFrame.Text(str, _) => str }
```

Comment:
I wonder if we're missing a whole class of web socket entity codecs. :thinking:

Reply:
What is a websocket entity codec about? Is there sth comparable in another ws client api?

Reply:
What is a web socket entity codec? I don't think I understand what it is supposed to do - "websocket interaction" seems to be to broad to be put into a useful codec. Are there other WS client APIs with a similar concept?

Reply:
It probably doesn't make sense as anything other than a pipe, on deeper thought. I was wondering about aggregating text frames into lines or a stream of JSON objects, like jawn's async mode.  I'm not sure if there are common pipes that everyone writes or not.  I'm commenting on a spec that I don't have a lot of practical experience with, so feel free to tell me I'm being silly.

Reply:
As you say, parsing a JSON stream is possible in two LOC due to the awesome `jawn-fs2`. But such abstractions could ofc be built on top of the WS API in the future, if they are considered necessary for other use cases I don't see right now.